File: backend/app/pool_selector.py
"""
Selección del pool efectivo de equipos según número de participantes.

Reglas:
1. Ordenar ``TEAMS`` por ``priority`` ASC.
2. Separar en clubes (``type == 'club'``) y selecciones (``type == 'nation'``).
3. Construir pool de tamaño ``N``:
   - Si ``N <= len(CLUBS)``: pool = primeros N clubes por priority.
   - Si ``N > len(CLUBS)``: pool = todos los clubes + selecciones por priority ASC.
"""
from __future__ import annotations

from typing import List, Tuple

from .teams import TEAMS


def split_pool() -> Tuple[List[dict], List[dict]]:
    """Devuelve ``(clubes, selecciones)`` ordenados por priority ASC."""
    sorted_teams = sorted((dict(t) for t in TEAMS), key=lambda t: t["priority"])
    clubs = [t for t in sorted_teams if t["type"] == "club"]
    nations = [t for t in sorted_teams if t["type"] == "nation"]
    return clubs, nations


def select_effective_pool(n: int) -> List[dict]:
    """
    Devuelve la lista de ``n`` equipos priorizando clubes sobre selecciones.

    :param n: cantidad de participantes (equipos a devolver).
    :raises ValueError: si ``n`` es inválido.
    """
    if n < 1:
        raise ValueError("n debe ser >= 1")
    if n > len(TEAMS):
        raise ValueError(f"n no puede exceder {len(TEAMS)} (pool total)")

    clubs, nations = split_pool()

    if n <= len(clubs):
        return [dict(t) for t in clubs[:n]]

    faltan = n - len(clubs)
    pool = [dict(t) for t in clubs] + [dict(t) for t in nations[:faltan]]
    return pool


def describe_pool(pool: List[dict]) -> dict:
    """Devuelve contadores de clubes y selecciones del pool dado."""
    clubs = sum(1 for t in pool if t["type"] == "club")
    nations = sum(1 for t in pool if t["type"] == "nation")
    return {"clubs_count": clubs, "nations_count": nations}

File: frontend/src/lib/poolSelector.ts
import type { Team } from "../api/client";
import { TEAMS } from "./teams";

export function splitPool(): { clubs: Team[]; nations: Team[] } {
  const sorted = [...TEAMS].sort((a, b) => a.priority - b.priority);
  return {
    clubs: sorted.filter((t) => t.type === "club"),
    nations: sorted.filter((t) => t.type === "nation"),
  };
}

export function selectEffectivePool(n: number): Team[] {
  if (n < 1) throw new Error("n debe ser >= 1");
  if (n > TEAMS.length) throw new Error(`n no puede exceder ${TEAMS.length} (pool total)`);
  const { clubs, nations } = splitPool();
  if (n <= clubs.length) return clubs.slice(0, n).map((t) => ({ ...t }));
  const faltan = n - clubs.length;
  return [...clubs.map((t) => ({ ...t })), ...nations.slice(0, faltan).map((t) => ({ ...t }))];
}

export function describePool(pool: Team[]): { clubs_count: number; nations_count: number } {
  return {
    clubs_count: pool.filter((t) => t.type === "club").length,
    nations_count: pool.filter((t) => t.type === "nation").length,
  };
}
